### DISCUSSION TITLE
Exercise 19:  Is safety violated? (Consensus)

### DISCUSSION THREAD

**Discussion by dag7dev** [sapienzastudentsnetwork/distributed-systems#44] [Exercises]
**Exercise 19:** Assume that all proposers are learners as well. Let even rounds be assigned to proposers with the rules that we know. Moreover, If round $2i$ is assigned to proposer $p$, then also round $2i+1$ is assigned to proposer $p$. Odd rounds are “recovery” rounds. If round $2i$ is a fast round and if the proposer of round $2i$ sees a conflict (it is also a learner), then the proposer immediately sends an accept for round $2i+1$ with the value that has been most voted in round $2i$, without any prepare and any promise. Is safety violated? If yes, show an example. If not, demonstrate safety.

**Comment by Tommaso-Sgroi** [comment]
It doesn’t violate safety. In Paxos, the safety theorem states that if an acceptor $a$ has voted for value $v$ in round $i$, then no value $v' \neq v$ can be chosen in rounds that are smaller than $i$. In Fast Paxos, when the coordinator recognizes a conflict in round $i$, it can always start a round that is greater than $i$ in order to solve it. It doesn’t break the second rule of safety, because the value $v$ will be accepted at a greater round.

**Comment by cornflexxx** [comment]
(Proof sketch) If there's no collision then the protocol is the same as Fast Paxos (which we know is safe). If there's some collision, a value is not chosen in the previous fast round, but no value is chosen in general ($\ast$). So in the next round the assigned proposer can send $\langle \text{ACCEPT}\rangle$ for the most voted value directly (actually the proposer can choose an arbitrary value since no value is chosen yet) $\implies$ safety holds.

$(\ast)$ No value is chosen in the fast round, and if we assume that a value was chosen in the previous round, we arrive at a contradiction, also assuming that the property holds by induction (base is easy), since more than one value is voted on during the fast round, but only one had been chosen previously.
